File: problems/0454.FourSumCount/FourSumCount.py
class Solution:
    def fourSumCount(self, nums1, nums2, nums3, nums4):
        """
        计算四个数组中各取一个元素，使得四元素之和为0的组合数量
        
        Args:
            nums1, nums2, nums3, nums4: 四个长度相同的整数数组
            
        Returns:
            int: 满足条件的四元组数量
        """
        # 使用字典存储前两个数组元素和的频率
        count_ab = {}
        for a in nums1:
            for b in nums2:
                count_ab[a + b] = count_ab.get(a + b, 0) + 1
        
        # 统计后两个数组元素和的相反数在字典中的出现次数
        result = 0
        for c in nums3:
            for d in nums4:
                target = -(c + d)
                if target in count_ab:
                    result += count_ab[target]
        
        return result
    
if __name__ == '__main__':
    s = Solution()
    print(s.fourSumCount([1, 2], [-2, -1], [-1, 2], [0, 2]))
    print(s.fourSumCount([0], [0], [0], [0]))

File: problems/0454.FourSumCount/FourSumCount.rs
use std::collections::HashMap;

struct Solution;

impl Solution {
    pub fn four_sum_count(nums1: Vec<i32>, nums2: Vec<i32>, nums3: Vec<i32>, nums4: Vec<i32>) -> i32 {
        // 使用哈希表存储前两个数组元素和的频率
        let mut count_ab = HashMap::new();
        for &a in &nums1 {
            for &b in &nums2 {
                *count_ab.entry(a + b).or_insert(0) += 1;
            }
        }
        
        // 统计后两个数组元素和的相反数在哈希表中的出现次数
        let mut result = 0;
        for &c in &nums3 {
            for &d in &nums4 {
                let target = -(c + d);
                if let Some(&count) = count_ab.get(&target) {
                    result += count;
                }
            }
        }
        
        result
    }
}

fn main() {
    let nums1 = vec![1, 2];
    let nums2 = vec![-2, -1];
    let nums3 = vec![-1, 2];
    let nums4 = vec![0, 2];
    
    let result = Solution::four_sum_count(nums1, nums2, nums3, nums4);
    println!("{}", result);
}